source/internal/array.d.ts

```diff
@@ -0,0 +1,98 @@
+import type {UnknownArray} from '../unknown-array';
+
+/**
+Infer the length of the given array `<T>`.
+
+@link https://itnext.io/implementing-arithmetic-within-typescripts-type-system-a1ef140a6f6f
+*/
+type ArrayLength<T extends readonly unknown[]> = T extends {readonly length: infer L} ? L : never;
+
+/**
+Matches any unknown array or tuple.
+*/
+export type UnknownArrayOrTuple = readonly [...unknown[]];
+// TODO: should unknown-array be updated?
```

Comment:
Do we need both of these types?

Reply:
I don't know. I don't think that's relevant to this pull request though. Open a separate issue about it.